### DISCUSSION TITLE
Wrong final balance

### DISCUSSION THREAD

**Discussion by mdavid217** [eprbell/rp2#31] [General]
Hi there.  I'm converting from manual spreadsheets to RP2 - really promising stuff!  I've just finished entering my BTC transactions dating back to 2014, but my final balances are off just slightly. I've summed the various pieces from the input data, and the sent balance is off just slightly. Any chance I can get your input? Not sure if user error or something else. Seems like something is getting double-counted.

**Comment by eprbell** [comment]
Nice! Typically such mismatches are due to missing or incorrectly entered transaction data. Have you already checked these FAQs?
* https://github.com/eprbell/rp2/blob/main/docs/user_faq.md#how-to-verify-that-i-entered-data-correctly-in-the-input-spreadsheet
* https://github.com/eprbell/rp2/blob/main/docs/user_faq.md#how-to-verify-that-tax-computation-is-correct

If the data has been entered correctly. the  "Account Balances" table in the tax sheet in the generated rp2_full_report.ods output file should match the amounts reported in your actual accounts.

**Reply by mdavid217** [reply]
Really appreciate the fast response! 

The data seem correct to me on the in-out form - not seeing values which might account for the difference. Specifically looking at the full report output, for the coinbase exchange, in the Tax tab, the acquire balance matches all the in-flows for coinbase in the in-out tab, the received balance matches the sum of items in intra-flow with to-exchange of coinbase, however, the sent balance in the tax tab exceeds the sum of out-flow from coinbase + intra-flow with the from exchange of Coinbase. (The tax tab shows a sent balance of 8.099 for coinbase, expecting 8.05. Final balance off by the corresponding difference.)

I sent you a PM on twitter if you'd like me to email you my raw data or I can PM you a dropboxs link. Any help gladly welcome, but I completely understand if you'd rather not dig into my specifics.

**Reply by mdavid217** [reply]
Wait. I found it. In the out section, it wants a crypto fee. However, the fee I pay is in USD. I was calcing that out to where the crypto sum equaled the actual fiat fee so that the dollar amounts in cap gains accounted for the fee. So, that is no longer a problem.

I was wondering about how to account for the fiat fee on sale, but I can adjust that by multiplying the net by the crypto total which adjusts for the fee on sale.

Thanks for essentially being a sounding board.. and again, thanks for all your efforts with RP2! This is going to make my life so much easier.

**Reply by eprbell** [reply]
Glad you found the problem and that RP2 is valuable to you! WRT how to represent fiat and crypto fees in RP2, you can express both. Check this FAQ: https://github.com/eprbell/rp2/blob/main/docs/user_faq.md#how-to-represent-fiat-vs-crypto-transaction-fees.

**Reply by mdavid217** [reply]
Ahh, good stuff. I must've looked past that! 

I am seeing some confusing results however. If I leave crypto field blank (which docs seem to indicate is acceptable) the process errors out with "rp2.rp2_error.RP2TypeError: Parameter 'crypto_fee' has non-RP2Decimal value None".  If I leave that set to 0 to avoid the error and also have the fiat_fee specified in the row[1] the USD Fee field in the full report's out-flow table indicates $0.00 / the USD out doesn't indicate the actual net proceeds. I did see it spit a warning, "WARNING: LINK OutTransaction (2021-12-13 19:43:29.662000+00:00, id 14): crypto_fee * spot_price != fiat_fee: 0.000000 != 2.344368"

I plan to look at it again with fresh eyes in the morning.

Again, thanks for all your help!

[1] I was sure to add "fiat_fee" to the out_header dict corresponding to the column I'm using.

**Reply by eprbell** [reply]
Curious: what is the name of the exchange that requires fiat_fee for out-transactions? And more specifically: what would happen if you don't have any fiat in the account and you want to sell some crypto? 

The reasoning behind what you're seeing (`None` crypto_fee not allowed in out-transactions, but allowed in in-transactions) has been discussed in #27. You can read the details in there, but the outcome of that discussion was that we don't know of any exchanges that require fiat fees in out transactions, so we assume crypto fee is always needed. However if you have an example that breaks this assumption, we can make crypto_fee and fiat_fee peers (they are already that way in in-transactions, because there are exchanges that accept both fee options in that case).

The warning you are seeing is just pointing out that the transaction has a fiat fee (2.344) that doesn't match crypto_fee (0): this is because, again, we assume crypto_fee in out-transactions to be always there. But if the assumption changes (as described above) we'll update the implementation and the logic will follow closely the one in in-transacitons (and there won't be a warning anymore).

**Reply by mdavid217** [reply]
Coinbase is an example. The cash is sourced from the proceeds of the sale. ( I don't think they allow you to do a sale where the proceeds are insufficient to cover the fee as far as the edge case of a $0 fiat balance on the account.) Sample transaction below. Technically this could be considered using the crypto for the fee, but in the transaction exports and from the interface, it is a fiat fee. This has also been my experience using fiat services with Crypto.com. I see the more common crypto fee using my GateIO account.

From discussion #27 you did mention, "Yes, that would be the third case I listed above: in case of value mismatch, RP2 will use the mismatched fiat_fee in its tax calculation, but will issue a warning." However, I don't see that being the case. It looks like it is ignoring the fiat_fee from what I can tell.  I don't mind putting a 0 in the crypto_fee field as long as it uses the fiat_fee column provided.

Here is a lightly formatted paste of a sale on coinbase. 

```
Sold Bitcoin −0.1250 BTC @ $41,124.40 per BTC
Price per coin $41,124.40
Payment Method USD Wallet
Coinbase Fee $76.59
Subtotal $5,140.55
Total $5,063.96
```

Having an out flow record with -.125 BTC, that spot price, 0 in crypto_fee, and 76.59 in fiat_fee shows the proceeds as the cost including the fee, so what I've been doing is marking down the spot price, so in that example, I used 40,511.68 to have the proceeds in the result match my expected values.

**Reply by eprbell** [reply]
Coinbase uses crypto fees (not fiat) for out-transactions: the USD fee you see reported by Coinbase is just the USD conversion of the crypto fee. As you say the fee comes out of the proceeds of the sale: what that means is that a part of the crypto you're selling is used to cover the fee. I have first-hand experience of this: in the past I sold all of one crypto while having 0 USD on Coinbase and the transaction went as I described above.

You may also want to check DaLI, which supports Coinbase and Coinbase Pro and generates RP2 input from your account using the Coinbase REST API.

Regarding the second part of your comment (on fiat fee being ignored): can you post a sample ODS file reproducing the issue? It's a lot easier to discuss based on a precise data set, rather than English.

**Reply by mdavid217** [reply]
Ahh. I wonder if the API provides that already broken out?  The tax reports and the normal reports for me show the gross value and the fiat figures - for me to split that into the sold vs. fee portions would definitely add some overhead to the process, again based on my input data. Either way will be giving DaLI a look as well.

As for fiat fee, I've uploaded a testcase to a repo here on github demonstrating the behavior I'm seeing.  https://github.com/mdavid217/rp2temp

**Reply by eprbell** [reply]
Thanks for the repro: will take a look at it tomorrow.

**Reply by eprbell** [reply]
I checked your repro: there is a bug that affects the report only (not tax calculation which is correct). In out-transaction if crypto_fee == 0 and fiat_fee > 0 the fiat_fee was reported as 0 in rp2_full_report (but it was there internally in computation of taxes, so results are not affected). I uploaded a new version with a fix.

In any case, I think the code is doing the right thing wrt fiat_fee and out-transactions. It allows you to specify crypto_fee == 0 and fiat_fee > 0 (because I don't know all exchanges and what they do), but it treats it as suspicious and issues a warning. The standard behavior for out-transactions should be to use a crypto_fee. The conclusion remains the same as in #27 : if we discover an exchange that allows fiat fees in out-transactions, then we should make both crypto and fiat fees optional in out-transactions.

**Reply by eprbell** [reply]
BTW, thanks for reporting the bug!

**Reply by mdavid217** [reply]
Not a problem. I pulled the latest version with pip and did see the fiat fee reflected in the output - however I'm not sure there wasn't a bug in the tax calc? It shows a cap gains of $1,950 vs. the $1,900 I'd expect. 

The actual USD out in the example case was $31,950 accounting for the sale of 1 BTC at $32,000 minus the $50 fiat fee.  In the input, you show USD Fee, USD In No Fee, and USD in With Fee. For the out-flow, if there were corresponding fields, they'd be like a USD Fee of $50, USD Out No Fee of $32,000, and a USD Out With (After) Fee of $31,950. The USD Out after fee value being the actual proceeds used in the capgains calc.

**Comment by eprbell** [comment]
The $32000 is the fiat value spent without the fee (see the Legend sheet and  the [code emitting the value](https://github.com/eprbell/rp2/blob/main/src/rp2/plugin/report/rp2_full_report.py#L483): the total spent in this case would be $32050, which after deducting the $50 fee becomes $32000. So I think cap gains of $1950 is correct: $32000 - $30050.

**Reply by mdavid217** [reply]
Minor note, on the out-flow, it'd be the fiat value received, not spent. Indeed that's the amount without fees. But again since it's received, the fees reduces the net proceeds rather than increasing it.

Thus,,cost basis is price + cost to acquire is $30,000 + 50, and proceeds from the sale is the gross sale receipt minus fees, $32,000 - 50.

I expect that to be $31,950 proceeds - $30,050 acquisition = $1,900 taxable cap gains. Right?

If there's a better way to express this in the spreadsheet than how I've done it, that's fine. Just trying to be 100% clear and get it right. Please know I'm not trying to be stubborn!

**Reply by eprbell** [reply]
No worries: it's all good, especially if you think there is a bug. It's possible I misinterpreted how to handle the out-transaction fee: not being an accountant, the code reflects my best understanding of tax regulations, but my understanding may not be 100% correct. So thanks for being stubborn :-) If you're correct RP2 is missing a small deduction during disposals.

This was already discussed in #6: back then the software would add the out fee to the the proceeds (so in this case capital gains would have been $32050 - $30050 = $2000). After that discussion I removed the fee from the calculation, so now we get $1950. If you're correct we should get $1900. However before making this change I want to be 100% sure: I will do a bit more research on this and if you have any pointers, please share. Again, thanks for nitpicking: my goal is to make RP2 as good as possible, so I appreciate people getting into the details and finding issues.

**Reply by mdavid217** [reply]
I can't understate how much I appreciate your professional and pragmatic attitude!

My understanding comes from when I started trading stocks before pretty much every major broker went commission free. I consider the fiat fees to be equivalent to the commissions brokerages used to charge for each individual trade. The standard practice with commissions was to factor them into the transactions - they would increase your cost basis and reduce your net proceeds. Resources online now don't talk about handling commission as widely as they used to due to their less-common prevalence. 

Here are some links which do include notes regarding commissions:
[Are Stock Commissions That Are Paid Tax Deductible on Your IRS Filing?](https://finance.zacks.com/stock-commissions-paid-tax-deductible-irs-filing-8230.html)
[Selling Stock: How Capital Gains are Taxed](https://www.fool.com/investing/how-to-invest/stocks/selling-stock-taxes/)
[Net Proceeds](https://www.investopedia.com/ask/answers/07/calculategains.asp)
[How Do I Calculate My Gains and Losses When I Sell a Stock?](https://www.investopedia.com/ask/answers/07/calculategains.asp)

**Comment by mdavid217** [comment]
I was looking into this last week. I found the changes that result in the output being what I'd expect to see (similar to issue #6).  The below patch would apply to out_transactions.py. 

The first change is within the init, making the fiat_out_with_fee subtract the fee rather than adding it. ("with fee" in this sense would mean "accounting for the fee" thus the subtraction of the fee from total fiat returned to the user from the transaction. The second change is to the fiat_taxable_amount property which would use the variable holding the fee rather than not. 

```
84c84
<         self.__fiat_out_with_fee = self.__fiat_out_no_fee + self.__fiat_fee
---
>         self.__fiat_out_with_fee = self.__fiat_out_no_fee - self.__fiat_fee
196c196
<         return self.fiat_out_no_fee
---
>         return self.fiat_out_with_fee
```

There are possibly a lot of ramifications from this change. For one thing, it breaks a ton of tests, although the output from a few of the test cases do look to correctly account for things. This change could result in savings on taxes by reducing reported capital gains by 1-2% depending on fees involved - maybe more if someone is paying a lot in gas fees relative to the size of their transactions. 

Another possible issue, is if someone hasn't been accounting for the fees previously (e.g. by adjusting down the spot price at time of sale to offset the difference) they would possibly be surprised / concerned / worried if they saw their cap gains from previous reports change. Additional, any user that *had* adjusted the transactions to account for this would need to know that they need to remove their adjustments if they want the new final gain/loss figures match previous output.

All the above considered, I think it would be best to leave the default behavior as-is. It's not technically wrong to essentially pay capgains tax on the fee, but it's not necessary either. Given that, probably best to avoid the headaches outlined above (and others I've probably not considered.)  However, it could be useful to have a configuration option / argument / global constant or something where users interested in it could opt in to account for fees as outlined above. 

Thoughts?

PS- is there any way to convert this discussion to an issue? I realized too late an issue would've been the better place for all of this.

**Comment by eprbell** [comment]
Thanks for following up, I haven't forgotten about this, but it's been on the backburner. I'm still trying to find official confirmation that this proposal is correct with crypto. The links you sent are interesting but they apply to stocks, whereas crypto is considered property for tax purposes. I still don't know for sure if stock and property are treated the same way when it comes to fee accounting: I would think so, but I'm no accountant. The closest I got to confirmation is this: https://www.irs.gov/publications/p544, especially "Example 1" in the "Gain or Loss From Sales and Exchanges" section, which shows that selling expenses are deducted from the sale price. However I'd like official confirmation from a CPA before making this change. I'm being a bit paranoid perhaps, but I really want to know this is correct before proceeding.

As for the change in results, I think it's OK: we could document this in the CHANGELOG.md file and respond to issues when people ask.

Not sure about how to convert to issue: if you find out how, let me know and I'll do it.
